pkg/packetfilter/packetfilter.go

```diff
@@ -0,0 +1,273 @@
+	ChainHookInput
+	ChainHookForward
+	ChainHookOutput
+	ChainHookPostrouting
+	ChainHookMAX
+)
+
+type ChainPriority uint32
+
+const (
+	ChainPriorityFirst ChainPriority = iota
+	ChainPriorityLast
+)
+
+type ChainType uint32
+
+const (
+	ChainTypeFilter ChainType = iota
+	ChainTypeRoute            // mangle
+	ChainTypeNAT
+	ChainTypeMAX
+)
+
+type ChainPolicy uint32
+
+const (
+	ChainPolicyAccept ChainType = iota
+	ChainPolicyDrop
+)
+
+type Rule struct {
+	DestCIDR string
+	SrcCIDR  string
+
+	SrcSetName  string
+	DestSetName string
+
+	SnatCIDR string
+	DnatCIDR string
+
+	OutIntreface string
+	TargetChain  string
+	MssValue     string
+
+	DPort     string
+	MarkValue string
+	Action    RuleAction
+	Proto     RuleProto
+	ClampType MssClampType
+}
+
+// Supported policy values are accept (which is the default) or drop.
+type ChainRegular struct {
+	Name   string
+	Policy ChainPolicy
+}
+
+type ChainIPHook struct {
+	Name     string
+	Type     ChainType
+	Hook     ChainHook
+	Priority ChainPriority
+	Policy   ChainPolicy
+	JumpRule *Rule
+}
+
+type Base interface {
+	// Chains
+	ChainExists(table TableType, chain string) (bool, error)
+	CreateIPHookChainIfNotExists(chain *ChainIPHook) error
+	CreateRegularChainIfNotExists(table TableType, chain *ChainRegular) error
+	DeleteIPHookChain(chain *ChainIPHook) error
+	DeleteRegularChain(table TableType, chain string) error
+	ClearChain(table TableType, chain string) error
+
+	// rules
+	Delete(table TableType, chain string, ruleSpec *Rule) error
+	InsertUnique(table TableType, chain string, position int, ruleSpec *Rule) error
+	PrependUnique(table TableType, chain string, ruleSpec *Rule) error
+	UpdateChainRules(table TableType, chain string, rules []*Rule) error
+	AppendUnique(table TableType, chain string, ruleSpec *Rule) error
+}
+
+type MssClampInterface interface {
+	SetMultiClusterMssClamp(clampType MssClampType, mssValue int) error
+	DelMultiClusterMssClamp() error
+	AddSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+	DelSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+}
+type GlobalNetEgressInterface interface {
+	CreateGlobalEgressForPods(key, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForPods(handle GlobalEgressHandle) error
+	CreateGlobalEgressForNamespace(namespace, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForNamespace(handle GlobalEgressHandle) error
+	AddPodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	RemovePodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+}
+
+type Driver interface {
+	Base
+	MssClampInterface
+	GlobalNetEgressInterface
+}
+type DriverCreateFunc func() (Driver, error)
+
+type driverInfo struct {
+	createFunc DriverCreateFunc
+	name       string
+	isDefault  bool
+}
+
+var (
+	logger                 = log.Logger{Logger: logf.Log.WithName("PacketFilter")}
+	NewFunc                func() (Base, error)
```

Comment:
I assume this is for unit tests to hook in like the current `iptables` abstraction but we don't need this  - the fake implementation can register with `AddDriver`.

Reply:
Thought to use it for unit testing hook, but yeah , I guess we can use AddDriver instead.